nomenklatura/resolver/resolver.py

```diff
@@ -167,145 +404,147 @@ def decide(
-            if node in self.nodes:
-                self.nodes[node].discard(edge)
+        stmt = update(self._table)
+        stmt = stmt.values({"deleted_at": timestamp()})
+        stmt = stmt.where(self._table.c.target == edge.target.id)
+        stmt = stmt.where(self._table.c.source == edge.source.id)
+        self._get_connection().execute(stmt)
 
     def _remove_node(self, node: Identifier) -> None:
         """Remove a node from the graph."""
-        edges = self.nodes.get(node)
-        if edges is None:
-            return
-        for edge in list(edges):
-            if edge.judgement != Judgement.NO_JUDGEMENT:
-                self._remove_edge(edge)
+        stmt = update(self._table)
+        stmt = stmt.values({"deleted_at": timestamp()})
+        cond = or_(
+            self._table.c.source == node.id,
+            self._table.c.target == node.id,
+        )
+        stmt = stmt.where(cond)
+        self._get_connection().execute(stmt)
 
     def remove(self, node_id: StrIdent) -> None:
         """Remove all edges linking to the given node from the graph."""
-        with self.lock:
-            node = Identifier.get(node_id)
-            self._remove_node(node)
-            self.connected.cache_clear()
+        node = Identifier.get(node_id)
+        self._remove_node(node)
+        self._invalidate()
 
     def explode(self, node_id: StrIdent) -> Set[str]:
         """Dissolve all edges linked to the cluster to which the node belongs.
         This is the hard way to make sure we re-do context once we realise
         there's been a mistake."""
-        with self.lock:
-            node = Identifier.get(node_id)
-            affected: Set[str] = set()
-            for part in self.connected(node):
-                affected.add(str(part))
-                self._remove_node(part)
-            self.connected.cache_clear()
-            return affected
+        node = Identifier.get(node_id)
+        affected: Set[str] = set()
+        for part in self.connected(node):
+            affected.add(str(part))
+            self._remove_node(part)
+        self._invalidate()
+        return affected
 
     def prune(self) -> None:
-        """Remove suggested (i.e. NO_JUDGEMENT) edges, keep only the n with the
-        highest score. This also checks if a transitive judgement has been
-        established in the mean time and removes those candidates."""
-        with self.lock:
-            for edge in list(self.edges.values()):
-                if edge.judgement == Judgement.NO_JUDGEMENT:
-                    self._remove_edge(edge)
-            self.connected.cache_clear()
-
-    def save(self) -> None:
+        """Remove suggested (i.e. NO_JUDGEMENT) edges."""
+        stmt = update(self._table)
+        stmt = stmt.where(self._table.c.judgement == Judgement.NO_JUDGEMENT.value)
+        stmt = stmt.where(self._table.c.deleted_at.is_(None))
+        stmt = stmt.values({"deleted_at": timestamp()})
```

Comment:
@pudo should we hard-delete NO_JUDGEMENT entries deleted, say, > 3 months ago? That'd mean they were candidates, we pruned them, and they haven't been a candidate again in a while.

Reply:
prune deletes all no judgements